source/rust_verify_test/tests/safe_api.rs

```diff
@@ -0,0 +1,231 @@
+        pub trait Foo {
+            fn test(x: u8) -> (y: u8)
+                no_unwind when x > 0;
+        }
+    } => Err(err) => assert_vir_error_msg(err, "Safe API violation: 'unwind' clause is nontrivial")
+}
+
+test_verify_one_file_with_options! {
+    #[test] trait_fn_mask_fail ["-V check-safe-api"] => verus_code! {
+        pub trait Foo {
+            fn test(x: u8) -> (y: u8)
+                opens_invariants none;
+        }
+    } => Err(err) => assert_vir_error_msg(err, "Safe API violation: 'invariant' clause is nontrivial")
+}
+
+test_verify_one_file_with_options! {
+    #[test] trait_fn_unsafe_ok ["-V check-safe-api"] => verus_code! {
+        pub unsafe trait Foo {
+            fn test(x: u8) -> (y: u8)
+                ensures y > 0;
+        }
+    } => Ok(())
+}
+
+test_verify_one_file_with_options! {
+    #[test] trait_fn_private_ok ["-V check-safe-api"] => verus_code! {
+        trait Foo {
+            fn test(x: u8) -> (y: u8)
+                ensures y > 0;
+        }
+    } => Ok(())
+}
+
+test_verify_one_file_with_options! {
+    #[test] trait_fn_impl_requires_fail ["-V check-safe-api"] => verus_code! {
+        pub trait Foo {
+            fn test(x: u8)
+                requires x > 0;
+        }
+
+        struct X { }
+
+        impl Foo for X {
+            fn test(x: u8) {
+            }
+        }
+    } => Err(err) => assert_vir_error_msg(err, "Safe API violation: 'requires' clause is nontrivial")
+}
+
+test_verify_one_file_with_options! {
+    #[test] trait_fn_impl_requires_fail_even_if_trait_is_unsafe ["-V check-safe-api"] => verus_code! {
+        pub unsafe trait Foo {
+            fn test(x: u8)
+                requires x > 0;
+        }
+
+        struct X { }
+
+        unsafe impl Foo for X {
+            fn test(x: u8) {
+            }
+        }
+    } => Err(err) => assert_vir_error_msg(err, "Safe API violation: 'requires' clause is nontrivial")
+}
+
+test_verify_one_file_with_options! {
+    #[test] trait_fn_impl_unsafe_ok ["-V check-safe-api"] => verus_code! {
+        pub trait Foo {
+            unsafe fn test(x: u8)
+                requires x > 0;
+        }
+
+        struct X { }
+
+        impl Foo for X {
+            unsafe fn test(x: u8) {
+            }
+        }
+    } => Ok(())
+}
+
+test_verify_one_file_with_options! {
+    #[test] trait_fn_impl_with_default_body_and_nontrivial_requires ["-V check-safe-api"] => verus_code! {
+        pub trait Foo {
+            fn test(x: u8)
+                requires x > 0
+            {
+            }
+        }
+
+        struct X { }
+
+        impl Foo for X {
+        }
+    } => Err(err) => assert_vir_error_msg(err, "Safe API violation: 'requires' clause is nontrivial")
+}
+
+// There are a few sneaky ways to smuggle private functions to the client
+// by using higher-order functions.
+// Right now, Verus doesn't support the relevant features; these tests
+// will catch these cases in the future.
```

Comment:
Should we add tests for `dyn Fn` and `Any`, too?

I'm assuming type_invariant is API-safe, because of the requirement that "a struct with a type invariant cannot have any fields public to the crate"; this might also make a good test case.